scrapy_po/middleware.py

```diff
@@ -20,9 +23,51 @@ class InjectionMiddleware:
     for this argument.
 
     XXX: should it really be a downloader middleware?
-    XXX: can this middleware allow to skip downloading
-    a page if it is not needed?
     """
+    def is_response_going_to_be_used(self, request, spider):
+        """Check whether the request's response is going to be used."""
+        callback = get_callback(request, spider)
+        plan, _ = build_plan(callback, {})
```

Comment:
👍 for making it work end-to-end, but it seems the current implementation is not complete.

There are 2 things to consider:

1) have user annotated response as DummyResponse in a callback?
2) does a dependency of any of the arguments need to access scrapy Response (e.g. page object asks for ReesponseData)?

Then I think actions should be these:

1. User haven't annotated response as DummyResponse. Always download response, pass it to the callback.
2. There is response: DummyResponse; page objects don't need response. Don't download; pass DummyResponse instance to the callback.
3. There is response: DummyResponse; page objects need a Response. Download a response, but **pass DummyResponse instance to the callback**, as user asked for that.

A next question is how you detect

1) if response is annotated as DummyResponse, and
2) if a page object needs a response.

Currently this is done by checking if there is an argument named `response`. I think this is not correct, both for callback and for page objects, but for different reasons.

Callback: Scrapy passes response as a positional argument, not as a keyword argument. It means the name of the argument doesn't matter: in case of `def parse(self, resp, page: ProductPage, response: DummyResponse):` response should be downloaded. 

Page objects / injectables: argument names and types doen't matter; what matters is an implementation of providers for the types which page object uses. Examples:
* There can be e.g. page object which receives `response` argument of type SplashResponse. 
* There can be a page object which receives `resp` argument of type ResponseData, and the provider for ResponseData needs an access to response. 
* There can be a page object which receives `response` argument of type ResponseData, but a provider which is currently registered for ResponseData doesn't use Scrapy at all (e.g. it gets a response from a local cache).

So, if I'm not mistaken, It means that the logic for callback and for page objects should be different:

1. For callback - check type of the first argument.
2. For page object - find providers for all dependencies, and check if any of the providers need a Response.

The tricky part here is "if any of the providers need a Response"; currently there is no way of doing it. I think we can use dependency injection again here, and inspect their `__init__` method; if Response or its subclass is found, then mark the provider as requiring a Response. This (or some other way of doing it - e.g. having an attribute `requires_scrapy_response = True` should be documented.

Reply:
I've just refactored the source code. Could you please take another review?